source/particles/utilities.cc

```diff
@@ -0,0 +1,221 @@
+      // Take care of components
+      const ComponentMask comps =
+        (space_comps.size() == 0 ? ComponentMask(fe.n_components(), true) :
+                                   space_comps);
+      AssertDimension(comps.size(), fe.n_components());
+
+      const auto n_comps = comps.n_selected_components();
+
+      // Global to local indices
+      std::vector<unsigned int> space_gtl(fe.n_components(),
+                                          numbers::invalid_unsigned_int);
+      for (unsigned int i = 0, j = 0; i < space_gtl.size(); ++i)
+        if (comps[i])
+          space_gtl[i] = j++;
+
+      // [TODO]: when the add_entries_local_to_global below will implement
+      // the version with the dof_mask, this should be uncommented.
+      // // Construct a dof_mask, used to distribute entries to the sparsity
+      // Table<2, bool> dof_mask(max_particles_per_cell * n_comps,
+      //                         fe.dofs_per_cell);
+      // dof_mask.fill(false);
+      // for (unsigned int i = 0; i < space_fe.dofs_per_cell; ++i)
+      //   {
+      //     const auto comp_i = space_fe.system_to_component_index(i).first;
+      //     if (space_gtl[comp_i] != numbers::invalid_unsigned_int)
+      //       for (unsigned int j = 0; j < max_particles_per_cell; ++j)
+      //         dof_mask(i, j * n_comps + space_gtl[comp_i]) = true;
+      //   }
+
+      std::vector<types::global_dof_index> dof_indices(fe.dofs_per_cell);
+      std::vector<types::particle_index>   particle_indices(
+        max_particles_per_cell * n_comps);
+
+      // FullMatrix<double> local_matrix(max_particles_per_cell,
+      // fe.dofs_per_cell);
+
+      while (particle != particle_handler.end())
+        {
+          const auto &cell = particle->get_surrounding_cell(tria);
+          const auto  dh_cell =
+            typename DoFHandler<dim, spacedim>::cell_iterator(*cell, &space_dh);
+          dh_cell->get_dof_indices(dof_indices);
+          const auto pic         = particle_handler.particles_in_cell(cell);
+          const auto n_particles = particle_handler.n_particles_in_cell(cell);
+          particle_indices.resize(n_particles * n_comps);
+          // local_matrix.reinit({n_particles, fe.dofs_per_cell});
+          Assert(pic.begin() == particle, ExcInternalError());
+          for (unsigned int i = 0; particle != pic.end(); ++particle, ++i)
+            {
+              const auto p_id = particle->get_id();
+              for (unsigned int j = 0; j < fe.dofs_per_cell; ++j)
+                {
+                  const auto comp_j =
+                    space_gtl[fe.system_to_component_index(j).first];
+                  if (comp_j != numbers::invalid_unsigned_int)
+                    constraints.add_entries_local_to_global(
+                      {p_id * n_comps + comp_j}, {dof_indices[j]}, sparsity);
+                }
+            }
+          // [TODO]: when this works, use this:
+          // constraints.add_entries_local_to_global(particle_indices,
```

Comment:
Why doesn't it work?

Reply:
In the rectangular case (i.e., when the row indices and the column indices are different, as in this case) this function calls throws an exception `ExcNotImplemented()`.